src/services/admin/AuditLogsService.ts

```diff
@@ -0,0 +1,500 @@
+                )
+            )
+          ),
+        ])
+      )
+      .andThen(([user, ...roles]) => {
+        const isUserNotAdminCollaborator = roles.some(
+          (role) =>
+            !role ||
+            (role !== CollaboratorRoles.Admin &&
+              role !== CollaboratorRoles.IsomerAdmin)
+        )
+
+        if (isUserNotAdminCollaborator) {
+          logger.warn(
+            `User ${email} is not an admin collaborator of all the given repos ${JSON.stringify(
+              repoNames
+            )}`
+          )
+
+          return errAsync(
+            new ForbiddenError(
+              "User is not an admin collaborator of all the given repos"
+            )
+          )
+        }
+
+        return okAsync(user)
+      })
+      .andThen((user) =>
+        // Step 3: Obtain a token for the user
+        ResultAsync.combine([
+          okAsync(user),
+          tokenServiceInstance.getAccessToken(),
+        ])
+      )
+      .andThen(([user, accessToken]) =>
+        // Step 4: Construct a fake user session data object and get audit logs
+        ResultAsync.combine(
+          repoNames.map((repoName) => {
+            const userSessionData = new UserWithSiteSessionData({
+              githubId: user.githubId,
+              accessToken,
+              isomerUserId: user.id.toString(),
+              email,
+              siteName: repoName,
+            })
+
+            return this.getAuditLogs(userSessionData, sinceDate, untilDate).map(
+              (auditLogs) => ({
+                siteName: repoName,
+                auditLogs,
+                snapshotTime: new Date(),
+              })
+            )
+          })
+        )
+      )
+      .andThen((auditLogDtos) => {
+        // Step 5: Prepare the audit log CSV files for each repo
+        const auditLogHeader = "Date,Time,Activity,User,Page,Remarks\n"
+
+        return ResultAsync.combine(
+          auditLogDtos.map(({ siteName, auditLogs, snapshotTime }) => {
+            const csvContent = auditLogs.reduce(
+              (csv, { timestamp, activity, actor, page, remarks }) => {
+                const recordDate = timestamp.toISOString().split("T")[0]
+                const recordHour = timestamp
+                  .getUTCHours()
+                  .toString()
+                  .padStart(2, "0")
+                const recordMinute = timestamp
+                  .getUTCMinutes()
+                  .toString()
+                  .padStart(2, "0")
+                const recordSecond = timestamp
+                  .getUTCSeconds()
+                  .toString()
+                  .padStart(2, "0")
+                const recordTime = `${recordHour}:${recordMinute}:${recordSecond} (UTC)`
+
+                return `${csv}${recordDate},${recordTime},${activity},"${actor}","${page}","${remarks}"\n`
+              },
+              auditLogHeader
+            )
+            const csvFileDate = snapshotTime.toISOString().split("T")[0]
```

Comment:
nit: use papaparse?

Reply:
No strong opinion on this, opted to use this way originally since the amount of logs is not very large. I think we can optimise this later if we find that this is running into performance issues?

Reply:
ill let you make the final call since non functional, but i feel that its not just pref issues (not really a concern since no real time req).

what i valued more was their out of the box handling for nity things like new lines. so in the future when you want to make this into a full fleded feature, you can easily parse and unparse heh

Reply:
Adjusted in 85d9b84f4f7a3953776fec96f23fba80f6ef790f.